### ISSUE TITLE
distributions.categorical() produces incorrect results on mps device

### ISSUE THREAD

**Issue by airza** [pytorch/pytorch#141457]
### 🐛 Describe the bug

I have a neural network that outputs a tensor of size (B,4,100), with the aim of producing Bx4 categorial distributions of length 100. After some experimentation, I have figured out that `torch.distributions.Categorical` produces different output depending on if this distribution is sampled on my Macbook's `mps` device or via the CPU. The `mps` device seems to always produce identical outputs when it is sampled.

I have created a [very small repository](https://github.com/airza/torch_bug) that reproduces the issue. Running `probs.py` imports and creates a distribution based on a tensor that is either running on the `mps` or `cpu` backend. For some reason, the `mps` background always produces an identical output for each run, while the `cpu` backend appears to sample correctly. The following is an example output from the `probs.py` script, which i hope is relatively self-explanatory:

entropy
tensor([[4.5363, 4.5379, 4.5646, 4.5854]], device='mps:0')
tensor([[4.5363, 4.5378, 4.5646, 4.5854]])
mpu samples
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
tensor([[24, 17, 23, 15]], device='mps:0')
cpu samples
tensor([[16, 99, 90, 16]])
tensor([[79, 94, 30, 82]])
tensor([[81, 94, 27, 49]])
tensor([[93, 13, 80,  2]])
tensor([[68, 41, 40, 37]])
tensor([[85, 67, 80, 27]])
tensor([[54, 35,  0, 65]])
tensor([[81, 42,  7, 88]])
tensor([[24, 61, 50, 11]])
tensor([[73, 67, 83, 72]])

I am not sure what is causing the problem. Am I correct in assuming this is a bug in the MPS backend rather than in my code somehow?

PS: believe my understanding of sampling a multidimensional `Categorical` distribution is correct: invoking `dist()` on a tensor of size M x N x P will normalize each row so that the sum along P is 1 and then sample independently to produce a tensor of size M x N. I would like to update the documentation so that this is more clear to others, but I wanted to verify this is correct first.

### Versions

PyTorch version: 2.5.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.4.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.0.40.1)
CMake version: version 3.27.3
Libc version: N/A

Python version: 3.12.0 (v3.12.0:0fb18b02c8, Oct  2 2023, 09:45:56) [Clang 13.0.0 (clang-1300.0.29.30)] (64-bit runtime)
Python platform: macOS-14.4.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] numpy==2.1.3
[pip3] torch==2.5.1
[conda] numpy                     1.24.2          py311h60f8152_0    conda-forge
[conda] pytorch                   2.3.1                  py3.11_0    pytorch
[conda] torchvision               0.18.1                py311_cpu    pytorch

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

**Comment by Isalia20** [comment]
`probs.py` loads a pickle file which might contain viruses. Best to post whatever is in that file in the .py file rightaway and I can take a look

**Comment by airza** [comment]
Hello, thanks for your quick response.
I tried to put in the hardcoded tensors that are stored by the `probs.pt` file directly into the example in that repository. But frustratingly, when i just include it as the hardcoded tensor the issue doesn't appear- both the mps and cpu tensors appropriately sample from the distribution.

My understanding is that the `weights_only` parameter in `torch.load` is used to safely load the weights of saved tensors without using `pickle`, so it _should_ be safe. Are you encountering a warning that it is not? I'm genuinely perplexed that the  tensor does not behave identically when copy-pasted from the REPL to a new variable, so let me know if there is anything I can do to help.

**Comment by airza** [comment]
After some testing on my own machine I have figured out the cause of the problem. For some reason, calling `permute` on a tensor, then using it to derive a Categorical distribution makes it always return the same sample. I have a _much_ simpler reproduction script now, which doesn't rely on the pickled tensors at all.

```
import torch
t1 = torch.Tensor([[.25,.1],[.25,.1],[.25,.1],[.25,.7]]).unsqueeze(2).permute(2,1,0)
t2 = t1.to('mps')
x = torch.distributions.Categorical(t1)
y = torch.distributions.Categorical(t2)
#samples normally, as expected
for i in range(10):
    print(x.sample())
#always returns [[2,2]]
for i in range(10):
    print(y.sample())
```

**Comment by Isalia20** [comment]
Thank you, very interesting. Not really a pytorch developer but I assume this won't be an easy fix. I'll try to look into it this week
